### ISSUE TITLE
There is not good way to catch "File not found" errors on static assets

### ISSUE THREAD

**Issue by tristan957** [nestjs/serve-static#139]
## I'm submitting a...

<pre><code>
[ ] Regression 
[ ] Bug report
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead post your question on Stack Overflow.
</code></pre>

## Current behavior

So if you setup the ServeStaticModule and provide serveStaticOptions: { fallthrough: false }, the module reports a 500 (Internal Server) error.

I don't know why this is not coming clearly to me, but I want to change the 500 error to a 404, because the file was not found.

I tried to use a global Internal Server filter and match the route in the case of 500 errors, but that wasn't possible because for some reason the filter doesn't get hit. I must be missing something higher level about NestJS.

## Expected behavior

I expect there to be an ergonomic way to handle errors when fallthrough is set to false.

## Minimal reproduction of the problem with instructions

Use the example in the nest repo: https://github.com/nestjs/nest/blob/master/sample/24-serve-static/src/app.module.ts#L8

and configure your ServeStaticModule to also serve your home directory (just an example) include `serveStaticOptions: { fallthrough: false }`. Then after you have set up the mount point, start the server and enter a URL under the mount point that would not exist in your home directory. Your browser should report a 500 error.

## What is the motivation / use case for changing the behavior?

It currently seems either impossible or not easily discoverable on how a developer can catch the serve-static 500 error and change it to a 404 for when a file is not found.

## Environment

<pre><code>
Nest version: 7.0.2

Honestly maybe this is an issue for the serve-static repo. I am unsure.

**Comment by kamilmysliwiec** [comment]
> It currently seems either impossible or not easily discoverable on how a developer can catch the serve-static 500 error and change it to a 404 for when a file is not found.

Have you tried creating a global exception filter?

**Comment by tristan957** [comment]
Yes. Sorry if that was not clear from the original issue. My exception filter aimed at catching these 500 errors in order to convert them to 404s.

The global exception filter does not get hit. I know this because my log statements did not come across the console. I can confirm that an instance of the exception filter was created, just not used. To create the global exception filter, I added a provider to my module using the documentation specified at docs.nestjs.com.

**Comment by kamilmysliwiec** [comment]
Please, provide a minimal repository which reproduces your issue.

**Comment by danielkbx** [comment]
Just ran into the same issue. Even if an exception filter would be working, I think, a 404 would be the correct HTTP response code.

**Comment by tristan957** [comment]
I'll see about getting a reproducible repo tonight.

**Comment by cyrilselasi** [comment]
@kamilmysliwiec Run into the same issue. Was able to catch the error and handle it appropriately here:

`@Catch()
export class ServeStaticExceptionFilter implements ExceptionFilter {

	catch( exception: any, host: ArgumentsHost ): any {

		const ctx = host.switchToHttp();
		const response = ctx.getResponse<Response>();
		const status = exception instanceof HttpException ? exception.getStatus() : HttpStatus.INTERNAL_SERVER_ERROR;

		if ( exception.code === "ENOENT" ) ) {
			response.sendStatus( HttpStatus.NOT_FOUND );

		}
		response.sendStatus( status );

	}
}`

Main gotcha for me was the Catch decorator, since the error thrown is not an HttpException, be sure to leave [catch everything](https://docs.nestjs.com/exception-filters#catch-everything).

**Comment by tristan957** [comment]
@cyrilselasi have you found a good way to use the filter only on the route where you are serving static content?

**Comment by cyrilselasi** [comment]
Not yet, I have it set as a global filter at the moment. That's working fine for my case

**Comment by micaelmbagira** [comment]
@kamilmysliwiec @tristan957 here is a repo that reproduces the issue https://github.com/micaelmbagira/nestjs-service-static-139.

**Comment by cedx** [comment]
Same issue here. Returning a 404 status code instead of catching the error 500 with a global exception filter is definitely the best option.

**Comment by kamilmysliwiec** [comment]
PRs are more than welcome (if anyone is interested) :)

**Comment by tristan957** [comment]
@kamilmysliwiec I think maybe all I will try later today. Does not seem too hard to add at least (if I understand the code correctly) for the Express side of things which I use.

**Comment by raphaelsoul** [comment]
Is it possible to disable index fallback to avoid this exception?

```
ServeStaticModule.forRoot({
      rootPath: path.join(__dirname, '../../../public'),
      serveStaticOptions: {
        index: false,
      },
    }),
```

```
[2021-03-25T10:35:11.080] [INFO] NestApplication - Nest application successfully started
[2021-03-25T10:35:13.787] [ERROR] ExceptionsHandler - ENOENT: no such file or directory, stat '/Users/raphaelsoul/projects/zuma/nodejs/mock-server/public/index.html' Error: ENOENT: no such file or directory, stat '/Users/raphaelsoul/projects/zuma/nodejs/mock-server/public/index.html'
[2021-03-25T10:35:13.788] [ERROR] ExceptionsHandler - ENOENT: no such file or directory, stat '/Users/raphaelsoul/projects/zuma/nodejs/mock-server/public/index.html' Error: ENOENT: no such file or directory, stat '/Users/raphaelsoul/projects/zuma/nodejs/mock-server/public/index.html'
```

**Comment by gdiasbruno** [comment]
> index fallback

I have same issue

**Comment by micalevisk** [comment]
> Is it possible to disable index fallback to avoid this exception?

@raphaelsoul I don't think so due to the following:

https://github.com/nestjs/serve-static/blob/72dc754f48d109215221339897a507d13388cd0b/lib/loaders/express.loader.ts#L27

https://github.com/nestjs/serve-static/blob/72dc754f48d109215221339897a507d13388cd0b/lib/loaders/abstract.loader.ts#L13-L15

**Comment by raphaelsoul** [comment]
https://github.com/nestjs/serve-static/blob/d0f4a89ca32a7d2315b03a045d7a57bc99f11771/lib/loaders/express.loader.ts#L44-L59

`res.sendFile` in `renderFn` causes `no such file` error.

I think it is the best practice. we dont need a global exeception filter
```
ServeStaticModule.forRoot({
      rootPath: resolve(__dirname, '../public'),
      serveRoot: '/',
}),
```
For SPAs it is never a problem cause you will always have an `index.html`(or using `index` specifying an filename)

For serving static file in traditional frontend project, just give a non-empty `serveRoot`.

Therefore, it seems there is no need for a PR to fix this, we just need update nestjs docs to declare usage for serving SPAs and non-SPAs
@kamilmysliwiec

**Comment by basz** [comment]
A other use case I have have is serving a static SPA where we need to inject data into the index.html. It needs to be routed through a controller. 

Currently i can have a controller that handles /index.html but /some-other-route would serve the index.html unparsed.

**Comment by Lokkve0126** [comment]
It is possible to use async await in nestjs fastify to solve this problem
```typescript
app.get(renderPath, async (req, res) => {
                    await new Promise((resolve, reject) => {
                        const stream = fs.createReadStream(indexFilePath);
                        stream.on("data", (chunk) => {
                            res.type('text/html').send(chunk);
                            resolve();
                        });
                        stream.on("error", () => {
                            reject(new NotFoundException()))
                        });
                    });
                });
```

**Comment by tuanpq1998** [comment]
I'm currently encountering a "404 File Not Found" error when using the HTTP custom exception filter. This poses an issue because we have static files. Ideally, we'd like to provide a more user-friendly experience than just a JSON response or header status code. Is there a way to implement a colorful custom 404 error page using static files in NestJS? I haven't been able to find a method within NestJS itself, so I resorted to serving static files through Nginx and configuring error handling there.

**Comment by drewB** [comment]
@tuanpq1998 you can do something like this:

``` not_found_exception_filter.ts 
import { ArgumentsHost, Catch, ExceptionFilter } from "@nestjs/common";
import { NextFunction } from "express";
import path, { join } from "path";
import { fileURLToPath } from "url";

@Catch()
export class NotFoundExceptionFilter implements ExceptionFilter {
    catch(exception : any, host: ArgumentsHost) {
        const ctx = host.switchToHttp();
        const response = ctx.getResponse();
        const next = ctx.getNext<NextFunction>();

        if ( exception.code === "ENOENT" )  {
            const __filename = fileURLToPath(import.meta.url);
            const __dirname = path.dirname(__filename);

            response.sendFile(join(__dirname, '..', '..', 'public', '404_page.html'));
        } else {
            next(exception);
        }
    }
}
```

And then in your app.module.ts add the provider:

```
    {
      provide: APP_FILTER,
      useClass: NotFoundExceptionFilter,
    }
```

So something like:

```
providers: [
    AppService,
    {
      provide: APP_FILTER,
      useClass: NotFoundExceptionFilter,
    }
  ],
  ```
  
  This will serve /public/404_page.html when a file is not found.

**Comment by kamilmysliwiec** [comment]
https://github.com/nestjs/serve-static/releases/tag/5.0.0
